File: ppt_gen/cache.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any


def file_hash(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()[:16]


def content_hash(*parts: Any) -> str:
    payload = json.dumps(parts, sort_keys=True, default=str)
    return hashlib.sha256(payload.encode()).hexdigest()[:16]


def cache_meta_path(cache_dir: Path, key: str) -> Path:
    return cache_dir / f"{key}.json"


def is_cache_valid(
    cache_dir: Path,
    key: str,
    *,
    expected_hash: str,
    output_path: Path,
) -> bool:
    meta = cache_meta_path(cache_dir, key)
    if not meta.exists() or not output_path.exists():
        return False
    stored = json.loads(meta.read_text())
    return stored.get("hash") == expected_hash


def write_cache_meta(cache_dir: Path, key: str, *, hash_value: str, output: Path) -> None:
    cache_dir.mkdir(parents=True, exist_ok=True)
    cache_meta_path(cache_dir, key).write_text(
        json.dumps({"hash": hash_value, "output": str(output)})
    )

File: lib/cache.mjs
import { createHash } from "node:crypto";
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

export function contentHash(...parts) {
  const payload = JSON.stringify(parts, (_key, value) =>
    typeof value === "bigint" ? value.toString() : value,
  );
  return createHash("sha256").update(payload).digest("hex").slice(0, 16);
}

function cacheMetaPath(cacheDir, key) {
  return join(cacheDir, `${key}.json`);
}

export async function isCacheValid(cacheDir, key, { expectedHash, outputPath }) {
  const meta = cacheMetaPath(cacheDir, key);
  if (!existsSync(meta) || !existsSync(outputPath)) return false;
  const stored = JSON.parse(await readFile(meta, "utf8"));
  return stored.hash === expectedHash;
}

export async function writeCacheMeta(cacheDir, key, { hashValue, output }) {
  await mkdir(cacheDir, { recursive: true });
  await writeFile(
    cacheMetaPath(cacheDir, key),
    JSON.stringify({ hash: hashValue, output: String(output) }),
  );
}
